### ISSUE TITLE
TypeError: null is not an object (evaluating 'a.apply') at instrumentMethod.js

### ISSUE THREAD

**Issue by kbgreenlight** [DataDog/browser-sdk#1345]
We're seeing this error in Safari mainly. I think the issue is in this [line](https://github.com/DataDog/browser-sdk/blob/main/packages/core/src/tools/instrumentMethod.ts#L12).

```
let instrumentation = instrumentationFactory(original)
```
If `original` is not a function then `instrumentation` is undefined. The error message from the title occurs when `instrumentation.apply` is called on this [line](https://github.com/DataDog/browser-sdk/blob/main/packages/core/src/tools/instrumentMethod.ts#L16).

That's my understanding. Not sure which functions are not working in Safari. I found the following calls in the code base so I'm guessing it's one of these:
```
instrumentMethodAndCallOriginal(window, ‘onerror’ ...
instrumentMethodAndCallOriginal(window, 'onunhandledrejection' ...
instrumentMethodAndCallOriginal(history, 'pushState' ...
instrumentMethodAndCallOriginal(history, 'replaceState' ...
instrumentMethodAndCallOriginal(XMLHttpRequest.prototype, 'open' ...
instrumentMethodAndCallOriginal(XMLHttpRequest.prototype, 'send' ...
instrumentMethodAndCallOriginal(XMLHttpRequest.prototype, 'abort' ...
```

Let me know if I can provide any more info.

**Comment by kbgreenlight** [comment]
```
(original) =>
      function () {
        const args = arguments as unknown as Parameters<OBJECT[METHOD]>
        let result

        if (before) {
          callMonitored(before, this as ThisParameterType<OBJECT[METHOD]>, args)
        }

        if (typeof original === 'function') {
          // eslint-disable-next-line @typescript-eslint/no-unsafe-call
          result = original.apply(this, args)
        }

        if (after) {
          callMonitored(after, this as ThisParameterType<OBJECT[METHOD]>, args)
        }

        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
        return result
      }
  )
```
I think ^^ that is `instrumentationFactory`. If `typeof original !== 'function'` then it looks like `result` will be undefined. Let me know if I'm misunderstanding. Here's that [condition](https://github.com/DataDog/browser-sdk/blob/main/packages/core/src/tools/instrumentMethod.ts#L54)

**Comment by bcaudan** [comment]
yes `result` will be undefined, but `instrumentation` is not `result`, `instrumentationFactory` is composed of two functions:

```
(original) =>  // factory
      function () { // actual instrumentation
        ...
        return result
      }
  )
```

**Comment by kbgreenlight** [comment]
Ah I see, you're right, I missed that inner function that's returned.

To answer your questions:
> are you able to reproduce it consistently?
We caught this error in sentry. It's happening to ~1 in 1000 total users so we aren't able to easily reproduce this.
Even if we tried to repro in mobile safari it still looks like it's just ~1 in maybe 20 or 30.

> does it happen on specific safari versions?
We've seen it in various safari versions 14.* and 15.* We've also seen it in chrome and mac, but 85% of our reported cases have come from mobile safari.

> can you share a public URL where the error has been observed? or provide a simpler setup that could demonstrate the error?
We don't have a public URL or an easy setup to demonstrate unfortunately. In our sentry stack traces this error seems to occur after xhr requests are made (to various endpoints, both internal and to 3rd parties)

**Comment by bcaudan** [comment]
Feel free to reopen if needed.

**Comment by karhidle** [comment]
We are also experiencing this issue. Using version `4.17.1` .
![Screen Shot 2022-08-29 at 10 03 44 AM](https://user-images.githubusercontent.com/45017812/187244838-fa5f9f24-da14-4457-a559-c242c68e6c48.png)

Same experience as @kbgreenlight described, errors happen after xhr requests. @bcaudan could you reopen this issue for further investigation?

**Comment by bcaudan** [comment]
Hello @karhidle,

it seems to be a duplicate of #1570 for which [a fix](#1697) will be available in the next release (hopefully this week).
You can follow #1570 to be notified about the resolution.
